h/services/bulk_annotation.py

```diff
@@ -0,0 +1,190 @@
+from typing import List, Union
+
+import sqlalchemy as sa
+from sqlalchemy.engine import Row
+from sqlalchemy.orm import Session
+from sqlalchemy.sql import Select
+
+from h.models import Annotation, AnnotationModeration, Group, User
+
+
+class BadDateFilter(ValueError):
+    """There is something wrong with the date filter provided."""
+
+
+class BadFieldSpec(ValueError):
+    """There is something wrong with the field you have specified."""
+
+
+def date_match(column, spec: dict):
+    """
+    Get an SQL comparator for a date column based on dict spec.
+
+    The dict can contain operators as keys and dates as values as per the
+    following complete (but nonsensical) filter:
+
+        {
+            "gt": "2012-11-30",
+            "gte": "2012-11-30",
+            "lt": "2012-11-30",
+            "lte": "2012-11-30",
+            "eq": "2012-11-30",
+            "ne": "2012-11-30",
+        }
+
+    :raises BadDateFilter: For unrecognised operators or no spec
+    """
+    if not spec:
+        raise BadDateFilter(f"No spec given to filter '{column}' on")
+
+    clauses = []
+
+    for op_key, value in spec.items():
+        if op_key == "gt":
+            clauses.append(column > value)
+        elif op_key == "gte":
+            clauses.append(column >= value)
+        elif op_key == "lt":
+            clauses.append(column < value)
+        elif op_key == "lte":
+            clauses.append(column <= value)
+        elif op_key == "eq":
+            clauses.append(column == value)
+        elif op_key == "ne":
+            clauses.append(column != value)
+        else:
+            raise BadDateFilter(f"Unknown date filter operator: {op_key}")
+
+    return sa.and_(*clauses)
```

Comment:
I've used a large if statements here in response to feedback about having to learn the dict from before. Now there is no dict to learn, each line is a stand alone. I've also added more comprehensive docs to show how to use it.

The intention here is to create a boundary for testing and thinking where we can assume date filtering works correctly, and then if we are curious dive in for the details.

Reply:
>  Yeah I think this is probably easier to read. I think the "query builder" pattern is a trade-off:

One thing I've realised it does which I think is quite useful is to force you into a style where each unitary behavior is more isolated. 

If you are doing one big query you can have parts of the query which relate to (for example) NIPSA behavior sprinkled and intermingled throughout other parts. In the style where you could put an `if` around each statement, it becomes more ordered and separated.